api_app/services/aad_authentication.py

```diff
@@ -278,44 +280,171 @@ def _get_users_inc_groups_from_response(self, users_graph_data, roles_graph_data
+                    user_roles = self._get_roles_for_principal(user_id, roles_graph_data, app_id_to_role_name)
+
                     if not any(user.id == user_id for user in users):
-                        users.append(User(id=user_id, name=user_name, email=user_email, roles=self._get_roles_for_principal(user_id, roles_graph_data, app_id_to_role_name)))
+                        users.append(AssignedUser(id=user_id, displayName=user_name, userPrincipalName=user_principal_name, roles=user_roles))
+                    else:
+                        user = next((user for user in users if user.id == user_id), None)
+                        user.roles = list(set(user.roles + user_roles))
 
             # Handle group endpoint response
             elif "directoryObjects" in user_data["body"]["@odata.context"]:
                 group_id = user_data["id"]
                 for group_member in user_data["body"]["value"]:
                     user_id = group_member["id"]
                     user_name = group_member["displayName"]
-                    user_email = group_member["mail"]
+                    user_principal_name = group_member["userPrincipalName"]
+
+                    group_roles = self._get_roles_for_principal(group_id, roles_graph_data, app_id_to_role_name)
 
                     if not any(user.id == user_id for user in users):
-                        users.append(User(id=user_id, name=user_name, email=user_email, roles=self._get_roles_for_principal(group_id, roles_graph_data, app_id_to_role_name)))
+                        users.append(AssignedUser(id=user_id, displayName=user_name, userPrincipalName=user_principal_name, roles=group_roles))
+                    else:
+                        user = next((user for user in users if user.id == user_id), None)
+                        user.roles = list(set(user.roles + group_roles))
 
         return users
 
-    def get_workspace_users(self, workspace: Workspace) -> List[User]:
+    def get_workspace_users(self, workspace: Workspace) -> List[AssignedUser]:
         msgraph_token = self._get_msgraph_token()
         sp_graph_data = self._get_app_sp_graph_data(workspace.properties["client_id"])
-        app_id_to_role_name = {app_role["id"]: app_role["value"] for app_role in sp_graph_data["value"][0]["appRoles"]}
-        roles_graph_data = self._get_user_role_assignments(workspace.properties["sp_id"], msgraph_token)
+        app_id_to_role_name = {app_role["id"]: (app_role["displayName"]) for app_role in sp_graph_data["value"][0]["appRoles"]}
+        roles_graph_data = self._get_user_role_assignments(workspace.properties["sp_id"])
         users_graph_data = self._get_user_details(roles_graph_data, msgraph_token)
         users_inc_groups = self._get_users_inc_groups_from_response(users_graph_data, roles_graph_data, app_id_to_role_name)
 
         return users_inc_groups
 
-    def get_workspace_user_emails_by_role_assignment(self, workspace: Workspace):
```

Comment:
@fortunkam where has this gone?

Reply:
investigating

Reply:
Looks like I was a little over-zealous in a code cleanup.  Will re-add and test.